morph.go

```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
 
 	var applied int
 	for i := 0; i < steps; i++ {
-		start := time.Now()
 		migrationName := sortedMigrations[i].Name
-		m.config.Logger.Println(formatProgress(fmt.Sprintf(migrationProgressStart, migrationName)))
-
-		down := downMigrations[migrationName]
-		if err := m.driver.Apply(down, true); err != nil {
+		if err := m.apply(downMigrations[migrationName], true); err != nil {
 			return applied, err
 		}
-
 		applied++
-		elapsed := time.Since(start)
-		m.config.Logger.Println(formatProgress(fmt.Sprintf(migrationProgressFinished, migrationName, fmt.Sprintf("%.4fs", elapsed.Seconds()))))
 	}
 
 	return applied, nil
 }
 
+// Diff returns the difference between the applied migrations and the available migrations.
+func (m *Morph) Diff(mode models.Direction) ([]*models.Migration, error) {
+	appliedMigrations, err := m.driver.AppliedMigrations()
+	if err != nil {
+		return nil, err
+	}
+
+	if mode == models.Down {
+		sortedMigrations := reverseSortMigrations(appliedMigrations)
+		downMigrations, err := findDownScripts(sortedMigrations, m.source.Migrations())
+		if err != nil {
+			return nil, err
+		}
+
+		diff := make([]*models.Migration, 0, len(downMigrations))
+		for i := 0; i < len(sortedMigrations); i++ {
+			diff = append(diff, downMigrations[sortedMigrations[i].Name])
+		}
+
+		return diff, nil
+	}
```

Comment:
I am trying to understand if we would ever need this? This just returns the down migrations for _all_ applied forward migrations. This is exactly the same code as `ApplyDown(-1)`. 

Whereas the forward diff actually calculates the diff between what's pending to be applied vs what's already applied

Reply:
>  This is exactly the same code as ApplyDown(-1)

Not if we limit the migrations. Let's say you want to rollback 3 migrations. You get the diff (in this case all of the applied), the you pick last 3 migrations, generate the plan accordingly and apply it.

Reply:
> You get the diff (in this case all of the applied),

But all of the applied will include all historical migrations as well right?

Reply:
That's correct. But that's the logical diff if you look at the opposite direction :p

Reply:
Right, so my point is - will this be helpful in real life? We are already getting the reverse migrations by creating the plan.

Reply:
Yes if you plan to dowgrade. You get the diff, you limit the diff and generate the plan. The other alternative is to ask for limit in this method. But I wanted to keep this simple.